### PR TITLE
[caffe2] Some small changes to InferBlobShapesAndTypes definition and SameAsInput Schema

### PR DESCRIPTION
**Pull request by Swetko**

### CODE REVIEW

**Comment by ezyang** [comment]
Segfaults in tests

**Comment by Swetko** [comment]
Thank you. Will look into that. I made those changes against the original Caffe2 repo before merging into pytorch and now after the merge it segfaults on my machine as well. I'll look into this!

**Comment by Swetko** [comment]
Are there any instructions how to run the tests locally on my machine?

**Comment by pjh5** [comment]
To run the python tests use https://github.com/pytorch/pytorch/blob/master/.jenkins/caffe2/test.sh#L114 . Just use 'python' instead of '$PYTHON' and '$CAFFE2_PYPATH' should be <path to your python's site-packages>/caffe2 (to find this you can run https://github.com/pytorch/pytorch/blob/master/.jenkins/caffe2/test.sh#L48 , or you can look at https://caffe2.ai/docs/faq.html#null__how-do-python-installations-work and learn what site-packages are, or you can just omit those lines and ignore those tests yourself). Or you can run the specific tests that are failing with pytest (don't remember the syntax though).

File: caffe2/python/pybind_state.cc [hunk 1]
```diff
@@ -1219,13 +1219,15 @@ void addGlobalMethods(py::module& m) {
 
         // Parse protobuffers to NetDefs
         std::vector<std::unique_ptr<caffe2::NetDef>> nets;
```
Comment (yinghai) [inline caffe2/python/pybind_state.cc]: Where is this variable used again?

File: caffe2/python/pybind_state.cc [hunk 2]
```diff
@@ -1237,13 +1239,15 @@ void addGlobalMethods(py::module& m) {
          const std::map<std::string, std::vector<TIndex>> blob_dimensions) {
         // Parse protobuffers to NetDefs
         std::vector<std::unique_ptr<caffe2::NetDef>> nets;
```
Comment (yinghai) [inline caffe2/python/pybind_state.cc]: Same here.

File: caffe2/python/pybind_state.cc (hunk 1 above)
Comment (Swetko) [reply caffe2/python/pybind_state.cc]: It is not used. It's used as a container for the unique_ptr to the nets. If the InferTensorShape... function has signature that takes vector<unique_ptr> then you cannot call it with existing structures, only newly created. I changed it to taking just vector<Netdef*> .

File: caffe2/python/pybind_state.cc (hunk 2 above)
Comment (Swetko) [reply caffe2/python/pybind_state.cc]: same reason. If you see another way of fixing the issue, please let me know!

**Comment by Swetko** [comment]
The test that I see failing is the C++ test   "test/fatal_signal_asan_no_sig_test". that one seems quite unrelated to my change. It crashes on my machine as well. Moreover when I do a fresh download of Pytorch master it crashes again on my machine.

File: caffe2/python/pybind_state.cc (hunk 1 above)
Comment (yinghai) [reply caffe2/python/pybind_state.cc]: Looking at the code, `nets` is useless and you can just remove it.

**Review by yinghai — COMMENTED** [review]
BTW, were you able to reproduce the segfault?

File: caffe2/core/operator_schema.cc [hunk 3]
```diff
@@ -228,9 +228,10 @@ OpSchema& OpSchema::IdenticalTypeAndShapeOfInputDim(int idx, int dim) {
 
 OpSchema& OpSchema::ScalarType(::caffe2::TensorProto_DataType dt) {
   return TensorInferenceFunction(
-      [dt](const OperatorDef&, const vector<TensorShape>& /*input_types*/) {
-        vector<TensorShape> out(1);
-        out[0].set_data_type(dt);
+      [dt](const OperatorDef& def, const vector<TensorShape>& /*input_types*/) {
+        TensorShape shape;
+        shape.set_data_type(dt);
+        vector<TensorShape> out(def.output_size(),shape);
```
Comment (yinghai) [inline caffe2/core/operator_schema.cc]: nit: space after `,`

File: caffe2/python/pybind_state.cc (hunk 1 above)
Comment (Swetko) [reply caffe2/python/pybind_state.cc]: Then I'd need to manually delete the created Nets?

Comment (yinghai) [reply caffe2/python/pybind_state.cc]: OK, you cannot. You need that to hold ownership of the net.

**Review by yinghai — CHANGES_REQUESTED** [review]
I cannot see a lot gain of changing the interface to raw pointer. But on the other hand, I understand the issue with `unique_ptr`. Maybe just use `shared_ptr`? I don't think performance is an issue here.

File: caffe2/python/pybind_state.cc (hunk 1 above)
Comment (Swetko) [reply caffe2/python/pybind_state.cc]: Yes that's what I meant.

**Comment by Swetko** [comment]
I'm sorry. I got confused and had to delete that branch and recreate the pull request.
What is gained by the using raw pointers is that the function being called does not need to have ownership of the objects. It just does an operation / inspection of them. It makes the interface and intent seem cleaner to me that way. 

As far as I understand,  unique_ptr   is used when you want to hold on to the object till you get out of scope/destroyed.  Shared_ptr is when multiple objects  want to hold on to an object.  A function that is performing an operation on an object does not need to hold on to the objects, provided they are being held by someone else. Anyways, either way is fine, but I thought it's cleaner intent with raw pointers.

**Comment by Swetko** [comment]
http://www.codingstandard.com/rule/8-2-4-do-not-pass-stdunique_ptr-by-const-reference/

**Pull request closed without merging**